### PR TITLE
Test, refactor and fix a bug in Basis.get_axis_angle

### LINKED ISSUE #63397: Basis.toAxisAngle return NaN as angle
### Godot version

4.0.dev #b3df275

### System information

MacOS 11.6

### Issue description

When the rotation angle is very small, the angle returned is NaN.

### Steps to reproduce

In c++:

```c++
Basis bugNan(1.00000024, 0 , 0.000100001693, 0, 1, 0, -0.000100009143, 0, 1.00000024);
bugNan.get_axis_angle(axis, angle);
print_line(angle) // NaN
```

### Minimal reproduction project

_No response_

### LINKED ISSUE #65179: Kinematic body's transform origin is set to nan by move_and_slide
### Godot version

v3.5.stable.official [991bb6ac7]

### System information

Win10, Linux (popOS 22.04), AMD, Nvidia, Intel, Ryzen

### Issue description

In our 3d game we discovered an elusive bug where kinematic bodies' interaction causes the transforms to be set to nan. The error happens right after calling move_and_slide and both the returned velocity and the kinematic body's transforms' origin have all nan members.
This makes further calculations impossible using move_and_slide().

The error occurs rarely and I could not find anything strange in the input to move_and_slide or in the transform of the objects before the call. 

I noticed it triggers mostly when kinematic bodies are dropped onto each other, so the sample project repeatedly creates this situation.

We tested it on a windows machine with nvidia gpu and amd processor and on a linux machine with intel cpu and amd gpu and the issue seems to be independent of the platform. 

### Steps to reproduce

I managed to create a minimal project that exhibits the error. The code is a bit convoluted as it is a cut down part of the more complex movement code from our game, but it is less than 100 lines. 

To reproduce the bug simply run the project and wait. You might have to wait a few minutes or restart the project to trigger the bug, but this was the most consistent I could make it happen.

When the bug occurs the code should halt at the breakpoint.

### Minimal reproduction project

[KinematicBodyBugHunt.zip](https://github.com/godotengine/godot/files/9467345/KinematicBodyBugHunt.zip)

### PR DESCRIPTION
**Pull request by fabriceci**
Follow up of #63190
Fix #63397
Fix #65179

I understand this code better now and I found where this code came from, there was a slight difference on the end of the algorithm, I used the original one, especially as it is used in many libraries such as three.js, it seems safer that way.

I cleaned up the code a bit and added some tests.

### CODE REVIEW

File: core/math/basis.cpp [hunk 1]
```diff
@@ -815,29 +815,29 @@ void Basis::get_axis_angle(Vector3 &r_axis, real_t &r_angle) const {
...
 		real_t zz = (rows[2][2] + 1) / 2;
-		real_t xy = (rows[1][0] + rows[0][1]) / 4;
-		real_t xz = (rows[2][0] + rows[0][2]) / 4;
-		real_t yz = (rows[2][1] + rows[1][2]) / 4;
-		if ((xx > yy) && (xx > zz)) { // rows[0][0] is the largest diagonal term
+		real_t xy = (rows[0][1] + rows[1][0]) / 4;
+		real_t xz = (rows[0][2] + rows[2][0]) / 4;
+		real_t yz = (rows[1][2] + rows[2][1]) / 4;
+
+		if ((xx > yy) && (xx > zz)) { // Rows[0][0] is the largest diagonal term.
```
Comment (akien-mga) [inline core/math/basis.cpp:840]: We have a rule to start comments with a capital letter, but here it's a variable identifier so it should be kept as in code (same below).
```suggestion
		if ((xx > yy) && (xx > zz)) { // rows[0][0] is the largest diagonal term.
```

File: core/math/basis.cpp [hunk 2]
```diff
@@ -815,29 +815,29 @@ void Basis::get_axis_angle(Vector3 &r_axis, real_t &r_angle) const {
...
-	if ((Math::abs(rows[1][0] - rows[0][1]) < CMP_EPSILON) && (Math::abs(rows[2][0] - rows[0][2]) < CMP_EPSILON) && (Math::abs(rows[2][1] - rows[1][2]) < CMP_EPSILON)) {
-		// singularity found
-		// first check for identity matrix which must have +1 for all terms
-		// in leading diagonal and zero in other terms
-		if ((Math::abs(rows[1][0] + rows[0][1]) < angle_epsilon) && (Math::abs(rows[2][0] + rows[0][2]) < angle_epsilon) && (Math::abs(rows[2][1] + rows[1][2]) < angle_epsilon) && (Math::abs(rows[0][0] + rows[1][1] + rows[2][2] - 3) < angle_epsilon)) {
-			// this singularity is identity matrix so angle = 0
+	*/
+
+	// http://www.euclideanspace.com/maths/geometry/rotations/conversions/matrixToAngle/index.htm
+	real_t x, y, z; // variables for result
```
Comment (akien-mga) [inline core/math/basis.cpp:821]:
```suggestion
	real_t x, y, z; // Variables for result.
```

File: core/math/basis.cpp [hunk 3]
```diff
@@ -815,29 +815,29 @@ void Basis::get_axis_angle(Vector3 &r_axis, real_t &r_angle) const {
...
-
-	if ((Math::abs(rows[1][0] - rows[0][1]) < CMP_EPSILON) && (Math::abs(rows[2][0] - rows[0][2]) < CMP_EPSILON) && (Math::abs(rows[2][1] - rows[1][2]) < CMP_EPSILON)) {
-		// singularity found
-		// first check for identity matrix which must have +1 for all terms
-		// in leading diagonal and zero in other terms
-		if ((Math::abs(rows[1][0] + rows[0][1]) < angle_epsilon) && (Math::abs(rows[2][0] + rows[0][2]) < angle_epsilon) && (Math::abs(rows[2][1] + rows[1][2]) < angle_epsilon) && (Math::abs(rows[0][0] + rows[1][1] + rows[2][2] - 3) < angle_epsilon)) {
-			// this singularity is identity matrix so angle = 0
+	*/
+
+	// http://www.euclideanspace.com/maths/geometry/rotations/conversions/matrixToAngle/index.htm
```
Comment (akien-mga) [inline core/math/basis.cpp:820]: Should use `https://` since it seems to work.

**Review by akien-mga — APPROVED** [review]

File: core/math/basis.cpp [hunk 4]
```diff
@@ -815,29 +815,29 @@ void Basis::get_axis_angle(Vector3 &r_axis, real_t &r_angle) const {
...
+	real_t x, y, z; // Variables for result.
+
+	if ((Math::abs(rows[0][1] - rows[1][0]) < CMP_EPSILON) && (Math::abs(rows[0][2] - rows[2][0]) < CMP_EPSILON) && (Math::abs(rows[1][2] - rows[2][1]) < CMP_EPSILON)) {
```
Comment (aaronfranke) [inline core/math/basis.cpp:823]: `abs(s) < CMP_EPSILON` is what `is_zero_approx` does, so we can just simplify it:

```suggestion
	if (Math::is_zero_approx(rows[0][1] - rows[1][0]) && Math::is_zero_approx(rows[0][2] - rows[2][0]) && Math::is_zero_approx(rows[1][2] - rows[2][1])) {
```

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
This would require a manual backport for 3.x, it's not trivial to cherry-pick.

**Pull request merged**